nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-hadoop-record-utils/src/main/java/org/apache/nifi/processors/hadoop/AbstractPutHDFSRecord.java

```diff
@@ -0,0 +1,505 @@
+
+                putFlowFile = postProcess(context, session, putFlowFile, destFile);
+
+                final String outputPath = destFile.toString();
+                final String newFilename = destFile.getName();
+                final String hdfsPath = destFile.getParent().toString();
+
+                // Update the filename and absolute path attributes
+                final Map<String,String> attributes = new HashMap<>(writeResult.get().getAttributes());
+                attributes.put(CoreAttributes.FILENAME.key(), newFilename);
+                attributes.put(ABSOLUTE_HDFS_PATH_ATTRIBUTE, hdfsPath);
+                attributes.put(RECORD_COUNT_ATTR, String.valueOf(writeResult.get().getRecordCount()));
+                putFlowFile = session.putAllAttributes(putFlowFile, attributes);
+
+                // Send a provenance event and transfer to success
+                final String transitUri = (outputPath.startsWith("/")) ? "hdfs:/" + outputPath : "hdfs://" + outputPath;
+                session.getProvenanceReporter().send(putFlowFile, transitUri);
+                session.transfer(putFlowFile, REL_SUCCESS);
+
+            } catch (IOException | FlowFileAccessException e) {
+                deleteQuietly(fileSystem, tempDotCopyFile);
+                getLogger().error("Failed to write due to {}", new Object[]{e});
+                session.transfer(session.penalize(putFlowFile), REL_RETRY);
+                context.yield();
+            } catch (Throwable t) {
+                deleteQuietly(fileSystem, tempDotCopyFile);
+                getLogger().error("Failed to write due to {}", new Object[]{t});
+                session.transfer(putFlowFile, REL_FAILURE);
+            }
+
+            return null;
+        });
+    }
+
+    /**
+     * This method will be called after successfully writing to the destination file and renaming the file to it's final name
+     * in order to give sub-classes a chance to take action before transferring to success.
+     *
+     * @param context the context
+     * @param session the session
+     * @param flowFile the flow file being processed
+     * @param destFile the destination file written to
+     * @return an updated FlowFile reference
+     */
+    protected FlowFile postProcess(final ProcessContext context, final ProcessSession session, final FlowFile flowFile, final Path destFile) {
+        return flowFile;
+    }
+
+    protected void rename(final FileSystem fileSystem, final Path srcFile, final Path destFile) throws IOException, InterruptedException, FailureException {
+        boolean renamed = false;
+        for (int i = 0; i < 10; i++) { // try to rename multiple times.
+            if (fileSystem.rename(srcFile, destFile)) {
+                renamed = true;
+                break;// rename was successful
+            }
+            Thread.sleep(200L);// try waiting to let whatever might cause rename failure to resolve
+        }
+        if (!renamed) {
```

Comment:
The behavior is that it will try up to 10 times to rename the dot file to the final name, and then if it still hasn't been renamed it will delete the dot file and route the flow file to failure. This behavior came from the existing PutHDFS so I kept the same behavior for consistency, but I can add something to the capability description of PutParquet describing this behavior.